devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -2,276 +2,481 @@
+   - Type: what temporal type to create (single datetime, range datetime)
+3. **Create `determine_temporal_spec()` function** that:
+   - Takes configuration, collection metadata, and available files
+   - Returns a complete specification of temporal decisions
+   - Encapsulates all business rules for temporal selection
+4. **Document decision rules** clearly:
+   - When to use single vs range datetime
+   - Collection temporal override logic
+   - Validation requirements
+5. **Add `TemporalSpec` to pipeline context** for use in later phases
+
+**Benefit if stopped here**: Temporal decision logic is isolated from execution, testable independently
+
+### Phase 4: Pipeline State Creation
+**Standalone Value**: Organized pipeline with clear data flow
+
+1. **Create `PipelineState`** class containing:
+   - Configuration
+   - CMR metadata (from Phase 1)
+   - Geometry specification (from Phase 2)
+   - Temporal specification (from Phase 3)
+   - Reader registry (empty for now)
+   - Processing ledger
+2. **Create new `Pipeline` class** that uses state
+3. **Update pipeline operations** to use specifications:
+   - Operations read the specs to know what to do
+   - Specs tell WHERE to get data and WHAT type to create
+   - Operations handle the HOW of extraction and formatting
+4. **Ensure all actions are pure functions** that return new state
+
+**Benefit if stopped here**: Clear separation between decision-making (specs) and execution (operations)
+
+### Phase 5: Reader Interface Definition
+**Standalone Value**: Standardized interface for all metadata extraction
+
+1. **Define new granular reader interfaces**:
+   - `TemporalReader`: Extracts temporal metadata
+   - `SpatialReader`: Extracts spatial metadata
+   - `AttributeReader`: Extracts general metadata/attributes
+2. **Create `ReaderRegistry`** to manage reader selection
+3. **Define reader selection logic** based on file types
+4. **Update pipeline state** to include reader registry
+
+**Benefit if stopped here**: Clear contract for all readers, foundation for extensibility
+
+### Phase 6: Custom Reader Implementation
+**Standalone Value**: Support for any file format without code changes
+
+1. **Implement custom reader base classes**:
+   - `CustomTemporalReader`
+   - `CustomSpatialReader`
+   - `CustomAttributeReader`
+2. **Add configuration parsing** for custom readers
+3. **Implement subprocess execution** with JSON output parsing
+4. **Add timeout and error handling**
+5. **Create example custom readers** and documentation
+6. **Register custom readers** in the reader registry
+
+**Benefit if stopped here**: Users can extend MetGenC to support any file format
+
+### Phase 7: Existing Reader Adaptation
+**Standalone Value**: All readers use consistent interface
+
+1. **Refactor `NetCDFReader`** to implement new interfaces:
+   - Split into temporal, spatial, and attribute readers
+2. **Refactor `CSVReader`** similarly
```

Comment:
As per discussion at our standup earlier in the week, I'm not sure we need to keep the CSV readers. Maybe the plan should break out "refactor netcdf reader" because that's the one that'll actually be used.